src/content/app/genome-browser/components/track-panel/components/track-panel-list/track-panel-items/TrackPanelGene.tsx

```diff
@@ -67,75 +59,76 @@ const TrackPanelGene = (props: TrackPanelGeneProps) => {
-
-  useEffect(() => {
-    const subscription = genomeBrowser?.subscribe(
-      IncomingActionType.VISIBLE_TRANSCRIPTS,
-      (action: ReportVisibleTranscriptsAction) => {
-        setVisibleTranscriptIds(action.payload.transcript_ids);
-      }
-    );
+  const { setFocusGene, updateFocusGeneTranscripts } = useGenomeBrowser();
+  const dispatch = useAppDispatch();
 
-    return () => subscription?.unsubscribe();
-  }, [genomeBrowser]);
+  if (!currentData) {
+    return null;
+  }
 
-  useEffect(() => {
-    updateObjectTrackStatus(trackStatus);
-  }, [genomeId, focusObjectId, trackStatus]);
+  const { gene } = currentData;
 
-  // set status of all transcripts based on the saved redux state after loading component
-  useEffect(() => {
-    if (allTranscriptsInGene?.length) {
+  const sortedTranscripts = defaultSort(gene.transcripts);
+  const visibleSortedTranscripts = isCollapsed
+    ? [sortedTranscripts[0]]
+    : sortedTranscripts;
+
+  const geneVisibilityStatus = !visibleTranscriptIds?.length
+    ? Status.UNSELECTED
+    : visibleTranscriptIds.length === gene.transcripts.length
+    ? Status.SELECTED
+    : Status.PARTIALLY_SELECTED;
+
+  const onGeneVisibilityChange = () => {
+    if (geneVisibilityStatus === Status.PARTIALLY_SELECTED) {
+      // show all transcripts
+      const visibleTranscriptIds = sortedTranscripts.map(
+        (transcript) => transcript.stable_id
+      );
       updateFocusGeneTranscripts(visibleTranscriptIds);
+      return;
     }
-  }, [allTranscriptsInGene]);
 
-  const setVisibleTranscriptIds = (transcriptIds: string[]) => {
-    dispatch(
-      updateObjectTrackStates({
-        status: transcriptIds.length ? Status.SELECTED : Status.UNSELECTED,
-        transcriptIds
-      })
-    );
-  };
+    const newStatus =
+      trackStatus === Status.SELECTED ? Status.UNSELECTED : Status.SELECTED;
 
-  const updateObjectTrackStatus = (newStatus?: Status) => {
-    if (!newStatus) {
-      newStatus =
-        trackStatus === Status.SELECTED ? Status.UNSELECTED : Status.SELECTED;
+    if (newStatus === Status.SELECTED) {
+      setFocusGene(focusObjectId);
+      const visibleTranscriptIds = sortedTranscripts.map(
+        (transcript) => transcript.stable_id
+      );
+      updateFocusGeneTranscripts(visibleTranscriptIds);
+    } else {
+      updateFocusGeneTranscripts([]);
     }
 
-    const newVisibleTranscriptIds =
-      newStatus === Status.SELECTED
-        ? allTranscriptsInGene?.map(({ stable_id }) => stable_id) ?? []
-        : [];
-
-    toggleTrack({ trackId: GENE_TRACK_ID, status: newStatus });
-    updateFocusGeneTranscripts(newVisibleTranscriptIds);
-
     dispatch(
       updateObjectTrackStates({
         status: newStatus
       })
     );
   };
 
-  if (!currentData) {
-    return null;
-  }
+  const onTranscriptVisibilityChange = (
+    transcriptId: string,
+    isVisible: boolean
+  ) => {
+    let updatedTranscriptIds = visibleTranscriptIds ?? ([] as string[]);
```

Comment:
This seems to be redundant as it gets overwritten two lines below

Reply:
I did this to get rid of the `null` that visibleTranscripts can be:

![image](https://user-images.githubusercontent.com/6834224/182401804-2a201fc7-cc9d-41c1-a349-12009b74e5d5.png)

![image](https://user-images.githubusercontent.com/6834224/182401879-1a82da86-c87a-4f8e-845d-6cb336726632.png)

Reply:
Seems like I was wrong to think we don't need to define it first at 117 separately. Let's keep it as it is.